src/api/c/canny.cpp

```diff
@@ -0,0 +1,248 @@
+/*******************************************************
+ * Copyright (c) 2017, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <af/dim4.hpp>
+#include <af/defines.h>
+#include <af/image.h>
+#include <af/seq.h>
+#include <err_common.hpp>
+#include <backend.hpp>
+#include <Array.hpp>
+#include <handle.hpp>
+#include <arith.hpp>
+#include <canny.hpp>
+#include <complex.hpp>
+#include <convolve.hpp>
+#include <copy.hpp>
+#include <histogram.hpp>
+#include <logic.hpp>
+#include <sobel.hpp>
+#include <transpose.hpp>
+#include <unary.hpp>
+#include <reduce.hpp>
+#include <ireduce.hpp>
+#include <index.hpp>
+#include <tile.hpp>
+#include <iota.hpp>
+#include <utility>
+#include <iostream>
+
+using af::dim4;
+using namespace detail;
+
+Array<float> gradientMagnitude(const Array<float>& gx, const Array<float>& gy, const bool& isf)
+{
+    if (isf) {
+        Array<float> gx2 = detail::abs<float, float>(gx);
+        Array<float> gy2 = detail::abs<float, float>(gy);
+        return detail::arithOp<float, af_add_t>(gx2, gy2, gx2.dims());
+    } else {
+        Array<float> gx2 = detail::arithOp<float, af_mul_t>(gx, gx, gx.dims());
+        Array<float> gy2 = detail::arithOp<float, af_mul_t>(gy, gy, gy.dims());
+        Array<float> sg  = detail::arithOp<float, af_add_t>(gx2, gy2, gx2.dims());
+        return detail::unaryOp<float, af_sqrt_t>(sg);
+    }
+}
+
+Array<float> otsuThreshold(const Array<float>& supEdges,
+                           const unsigned NUM_BINS, const float maxVal)
+{
+    Array<uint> hist = detail::histogram<float, uint, false>(supEdges, NUM_BINS, 0, maxVal);
+
+    const af::dim4 hDims = hist.dims();
+
+    // reduce along histogram dimension i.e. 0th dimension
+    auto totals  = reduce<af_add_t, uint, float>(hist, 0);
+
+    // tile histogram total along 0th dimension
+    auto ttotals = tile(totals, af::dim4(hDims[0]));
+
+    // pixel frequency probabilities
+    auto probability = arithOp<float, af_div_t>(cast<float, uint>(hist), ttotals, hDims);
+
+    af_index_t seqBegin[4];
+    af_index_t seqRest[4];
+
+    seqBegin[0].idx.seq = af_make_seq(0, hDims[0]-1, 1);
+    seqBegin[0].isSeq   = true;
+    seqBegin[0].isBatch = false;
+    seqRest[0].idx.seq = af_make_seq(0, hDims[0]-1, 1);
+    seqRest[0].isSeq   = true;
+    seqRest[0].isBatch = false;
+
+    for (int s=1; s<4; ++s)
+    {
+        seqBegin[s].idx.seq = af_span;
+        seqBegin[s].isSeq   = true;
+        seqBegin[s].isBatch = false;
+
+        seqRest[s].idx.seq = af_span;
+        seqRest[s].isSeq   = true;
+        seqRest[s].isBatch = false;
+    }
+
```

Comment:
None of this is required, You are only using sequences! No reason to store it in `af_index_t`